algorithms/src/polycommit/kzg10/data_structures.rs

```diff
@@ -206,6 +206,61 @@ pub struct VerifierKey<E: PairingEngine> {
     pub prepared_beta_h: <E::G2Affine as PairingCurve>::Prepared,
 }
 
+impl<E: PairingEngine> CanonicalSerialize for VerifierKey<E> {
+    fn serialize_with_mode<W: Write>(&self, mut writer: W, compress: Compress) -> Result<(), SerializationError> {
+        self.g.serialize_with_mode(&mut writer, compress)?;
+        self.gamma_g.serialize_with_mode(&mut writer, compress)?;
+        self.h.serialize_with_mode(&mut writer, compress)?;
+        self.beta_h.serialize_with_mode(&mut writer, compress)?;
+        Ok(())
+    }
+
+    fn serialized_size(&self, compress: Compress) -> usize {
+        self.g.serialized_size(compress)
+            + self.gamma_g.serialized_size(compress)
+            + self.h.serialized_size(compress)
+            + self.beta_h.serialized_size(compress)
+    }
+}
+
+impl<E: PairingEngine> CanonicalDeserialize for VerifierKey<E> {
+    fn deserialize_with_mode<R: Read>(
+        mut reader: R,
+        compress: Compress,
+        validate: Validate,
+    ) -> Result<Self, SerializationError> {
+        let g = CanonicalDeserialize::deserialize_with_mode(&mut reader, compress, validate)?;
+        let gamma_g = CanonicalDeserialize::deserialize_with_mode(&mut reader, compress, validate)?;
+        let h: E::G2Affine = CanonicalDeserialize::deserialize_with_mode(&mut reader, compress, validate)?;
+        let beta_h: E::G2Affine = CanonicalDeserialize::deserialize_with_mode(&mut reader, compress, validate)?;
+        let prepared_h = h.prepare();
+        let prepared_beta_h = beta_h.prepare();
+        Ok(VerifierKey { g, gamma_g, h, beta_h, prepared_h, prepared_beta_h })
+    }
+}
+
+impl<E: PairingEngine> Valid for VerifierKey<E> {
+    fn check(&self) -> Result<(), SerializationError> {
+        Valid::check(&self.g)?;
+        Valid::check(&self.gamma_g)?;
+        Valid::check(&self.h)?;
+        Valid::check(&self.beta_h)?;
+        Ok(())
+    }
+
+    fn batch_check<'a>(batch: impl Iterator<Item = &'a Self> + Send) -> Result<(), SerializationError>
```

Comment:
```suggestion
    fn batch_check(batch: Vec<Self>) -> Result<(), SerializationError>
```
if we're immediately going to allocate into a vector, it would be best to make it clear to the caller by expecting a vector to begin with

Reply:
Unfortunately sometimes the input actually is an iterator (e.g. `BTreeMap::iter`)

Reply:
it could still be collected into a `Vec` beforehand; it's only to not conceal possibly large allocations from the caller, unless that's not a concern here

Reply:
This would also require changing the trait definition, which is out-of-scope for this PR, but something we should consider in a future PR.